File: exdir-browser/exdirbrowser/models/exdirtreemodel.cpp
#include "exdirtreemodel.h"

#include <elegant-exdir/File>

#include <QDebug>

#include <QElapsedTimer>
#include <QFile>
#include <QFileInfo>
#include <QQmlFile>

using namespace elegant::exdir;

ExdirTreeModel::ExdirTreeModel(QObject *parent)
    : QAbstractItemModel(parent)
{
}

ExdirTreeModel::~ExdirTreeModel()
{
    if(m_root) {
        delete m_root;
    }
}

QModelIndex ExdirTreeModel::index(int row, int column, const QModelIndex &parent) const
{
    if(!hasIndex(row, column, parent)) {
        return QModelIndex();
    }

    const ExdirTreeItem* parentNode = nullptr;
    if(!parent.isValid()) {
        if(m_root == nullptr) {
            return QModelIndex();
        }
        parentNode = m_root;
    } else {
        parentNode = item(parent);
    }

    ExdirTreeItem* childItem = parentNode->child(row);
    if(childItem) {
        return createIndex(row, column, childItem);
    } else {
        return QModelIndex();
    }
}

QModelIndex ExdirTreeModel::parent(const QModelIndex &index) const
{
    if(!index.isValid()) {
        return QModelIndex();
    }

    ExdirTreeItem* childItem = item(index);
    ExdirTreeItem* parentItem = childItem->parentItem;

    if(parentItem == m_root) {
        return QModelIndex();
    }
    return createIndex(parentItem->row, 0, parentItem);
}

int ExdirTreeModel::rowCount(const QModelIndex &parent) const
{
    if(parent.column() > 0) {
        return 0;
    }
    if(!parent.isValid()) {
        if(m_root == nullptr) {
            return 0;
        }
        return m_root->children().count();
    }
    ExdirTreeItem* parentItem = item(parent);
    if(parentItem->needsChildIteration) {
        QString filePath = QQmlFile::urlToLocalFileOrQrc(m_source);
        File file(filePath.toStdString(), File::OpenMode::ReadWrite);
        Object object = file[parentItem->path().toStdString()];
        if(object.isGroup()) {
            Group group = object;
            addChildObjects(parentItem, group, parentItem->depth + 1);
        }
        parentItem->needsChildIteration = false;
    }
    return parentItem->children().count();;
}

int ExdirTreeModel::columnCount(const QModelIndex &parent) const
{
    if(parent.column() > 0) {
        return 0;
    }
    return 1;
}

QVariant ExdirTreeModel::data(const QModelIndex &index, int role) const
{
    if(!index.isValid()) {
        return QVariant();
    }
    ExdirTreeItem* indexNode = item(index);
    if(!indexNode) {
        return QVariant();
    }
    QString value;
    switch(role) {
    case int(Role::Name):
        value = indexNode->name();
        break;
    case int(Role::Path):
        value = indexNode->path();
        break;
    case int(Role::Type):
        value = indexNode->type();
        break;
    default:
        value = "";
        break;
    }
    return QVariant::fromValue(value);
}

QHash<int, QByteArray> ExdirTreeModel::roleNames() const
{
    QHash<int, QByteArray> roles;
    roles[int(Role::Name)] = "name";
    roles[int(Role::Path)] = "path";
    roles[int(Role::Type)] = "type";
    return roles;
}

QString ExdirTreeModel::path(const QModelIndex &index) const
{
    if(!index.isValid()) {
        return "";
    }
    ExdirTreeItem* indexNode = item(index);
    return indexNode->m_path;
}

ExdirTreeItem *ExdirTreeModel::item(const QModelIndex &index) const
{
    if(!index.isValid()) {
        return nullptr;
    }
    return static_cast<ExdirTreeItem*>(index.internalPointer());
}

QUrl ExdirTreeModel::source() const
{
    return m_source;
}

void ExdirTreeModel::setSource(QUrl source)
{
    if (m_source == source)
        return;

    m_source = source;
    loadFile();
    emit sourceChanged(source);
}

void ExdirTreeModel::addChildObjects(ExdirTreeItem* parent, const Group& parentGroup, int depth) const
{
    int row = 0;
    for(const std::string& key : parentGroup.keys()) {
        Object object = parentGroup[key];

        QString type = "";
        QString info = "";
        switch(object.type()) {
        case Object::Type::Attribute:
            type = "Attribute";
            break;
        case Object::Type::Dataset:
        {
            type = "Dataset";
            Dataset dataset = object;
            if(dataset.dimensionCount() == 1) {
                info = QString("vector of size %1").arg(dataset.extents()[0]);
            } else if(dataset.dimensionCount() == 2) {
                info = QString("%1x%2 matrix")
                        .arg(dataset.extents()[0])
                        .arg(dataset.extents()[1]);
            } else if(dataset.dimensionCount() == 3) {
                info = QString("%1x%2x%3 cube")
                        .arg(dataset.extents()[0])
                        .arg(dataset.extents()[1])
                        .arg(dataset.extents()[2]);
            } else {
                info = QString("%1 dimensional object").arg(dataset.dimensionCount());
            }
            break;
        }
        case Object::Type::Dataspace:
            type = "Dataspace";
            break;
        case Object::Type::Datatype:
            type = "Datatype";
            break;
        case Object::Type::File:
            type = "File";
            break;
        case Object::Type::Group:
        {
            type = "Group";
            Group group = object;
            info = QString("%1 objects").arg(group.keys().size());
            break;
        }
        default:
            qDebug() << "Got unknown type";
            type = "Unknown type";
            break;
        }

        ExdirTreeItem* node = new ExdirTreeItem(row, 0, depth + 1,
                                            QString::fromStdString(key),
                                            parent->m_path + "/" + QString::fromStdString(key),
                                            type,
                                            parent);
        node->setInfo(info);

        if(object.isGroup()) {
            node->needsChildIteration = true;
        }
        row += 1;
    }
}

void ExdirTreeModel::loadFile()
{
    QElapsedTimer timer;
    timer.start();
    qDebug() << "Loading tree";
    if(!m_source.isValid() || m_source.isEmpty()) {
        qDebug() << "Not loading because" << m_source;
        return;
    }
    QString filePath = QQmlFile::urlToLocalFileOrQrc(m_source);
    QString filenameOnly = filePath;
    QFileInfo fileInfo(filePath);
    if(fileInfo.exists()) {
        filenameOnly = fileInfo.fileName();
    }
    File file(filePath.toStdString(), File::OpenMode::ReadWrite);
    if(m_root) {
        delete m_root;
    }
    m_root = new ExdirTreeItem(0, 0, 0, "", "", "", 0);
    beginInsertRows(QModelIndex(), 0, 0);
    ExdirTreeItem *fileItem = new ExdirTreeItem(0, 0, 1, filenameOnly, "", "File", m_root);
    endInsertRows();
    addChildObjects(fileItem, file, 0);
    emit dataChanged(QModelIndex(), QModelIndex());

    qDebug() << "Done loading tree:" << timer.elapsed();
}

ExdirTreeItem::ExdirTreeItem() {}

ExdirTreeItem::ExdirTreeItem(int row_, int column_, int depth_, QString name_, QString path_, QString type, ExdirTreeItem *parent)
    : QObject(parent)
    , row(row_)
    , column(column_)
    , depth(depth_)
    , m_name(name_)
    , m_path(path_)
    , m_type(type)
    , parentItem(parent) {}

ExdirTreeItem::~ExdirTreeItem() {
    qDeleteAll(children());
}

ExdirTreeItem *ExdirTreeItem::child(int row) const {
    return static_cast<ExdirTreeItem*>(children().at(row));
}

QString ExdirTreeItem::name() const
{
    return m_name;
}

QString ExdirTreeItem::path() const
{
    return m_path;
}

QString ExdirTreeItem::type() const
{
    return m_type;
}

QString ExdirTreeItem::info() const
{
    return m_info;
}

void ExdirTreeItem::setName(QString name)
{
    if (m_name == name)
        return;

    m_name = name;
    emit nameChanged(name);
}

void ExdirTreeItem::setPath(QString path)
{
    if (m_path == path)
        return;

    m_path = path;
    emit pathChanged(path);
}

void ExdirTreeItem::setType(QString type)
{
    if (m_type == type)
        return;

    m_type = type;
    emit typeChanged(type);
}

void ExdirTreeItem::setInfo(QString info)
{
    if (m_info == info)
        return;

    m_info = info;
    emit infoChanged(info);
}

File: exdir-browser/exdirbrowser/models/exdirtreemodel.py
import exdir
from PyQt5.QtCore import *
from PyQt5.QtQml import *

class ExdirTreeModel(QAbstractItemModel):
    class Role:
        Name = Qt.UserRole + 0
        Path = Qt.UserRole + 1
        Type = Qt.UserRole + 2
    Q_ENUMS(Role)
    
    def __init__(self, parent):
        super().__init__(parent)
        self._source = None
        self.root = None

    def index(self, row, column, parent):
        if not self.hasIndex(row, column, parent):
            return QModelIndex()
        parentNode = None
        if not parent.isValid():
            if self.root is None:
                return QModelIndex()
            parentNode = self.root
        else:
            parentNode = self.item(parent)

        childItem = parentNode.child(row)
        if childItem:
            index = self.createIndex(row, column, childItem)
            # index = QPersistentModelIndex(self.createIndex(row, column, childItem))
            # index = QModelIndex(row, column, childItem, self)
            return index
        else:
            return QModelIndex()
    
    @pyqtSlot(QAbstractItemModel, QAbstractItemModel, result=str)
    def check(self, model1, model2):
        return "blah"

    def parent(self, index):
        if not index.isValid():
            return QModelIndex()
            
        childItem = self.item(index)
        parentItem = childItem.parentItem

        if parentItem == self.root:
            return QModelIndex()

        return self.createIndex(parentItem.row, 0, parentItem)

    def rowCount(self, parent):
        if parent.column() > 0:
            return 0

        if not parent.isValid():
            if self.root == None:
                return 0
            return len(self.root.children())

        parentItem = self.item(parent)
        if parentItem.needsChildIteration:
            filePath = self.source.toLocalFile()
            # TODO replace with Python exdir file
            file = exdir.File(filePath)
            object = file[parentItem.path]
            if isinstance(object, exdir.core.Group):
                group = object
                self.addChildObjects(parentItem, group, parentItem.depth + 1)
            parentItem.needsChildIteration = False
        return len(parentItem.children())

    def columnCount(self, parent):
        if parent.column() > 0:
            return 0
        return 1
    
    def data(self, index, role):
        if not index.isValid():     
            return QVariant()

        indexNode = self.item(index)
        if not indexNode:        
            return QVariant()

        value = {
            self.Role.Name: indexNode.name,
            self.Role.Path: indexNode.path,
            self.Role.Type: indexNode.type
        }.get(role, "")

        return QVariant(value)

    def roleNames(self):
        roles = {
            self.Role.Name: b"name",
            self.Role.Path: b"path",
            self.Role.Type: b"type",
        }
        return roles

    def path(self, index):
        if not index.isValid():        
            return ""

        indexNode = self.item(index)
        return indexNode.path
        
    @pyqtSlot(QModelIndex, result=QObject)
    def item(self, index):
        if not index.isValid():
            return None
        return index.internalPointer()

    def source(self):
        return self._source

    def setSource(self, source):
        if self._source == source:
            return

        self._source = source
        self.loadFile()
        self.sourceChanged.emit(source)

    def addChildObjects(self, parent, parentGroup, depth):
        row = 0
        for key in parentGroup.keys():
            object = parentGroup[key]

            type = ""
            info = ""

            # TODO needs major rewrite
            # if isinstance(object, exdir.Attribute):
                # type = "Attribute"
            if isinstance(object, exdir.core.Dataset):
                type = "Dataset"
            # elif isinstance(object, exdir.core.Datatype):
                # type = "Datatype"
            elif isinstance(object, exdir.core.File):
                type = "File"
            elif isinstance(object, exdir.core.Group):
                type = "Group"
                
            # group = object
            # info = QString("%1 objects").arg(group.keys().size())
                
            # dataset = object
            # if dataset.dimensionCount() == 1:                
            #     info = QString("vector of size %1").arg(dataset.extents()[0])
            # elif dataset.dimensionCount() == 2:
            #     info = QString("%1x%2 matrix")
            #             .arg(dataset.extents()[0])
            #             .arg(dataset.extents()[1])
            # elif dataset.dimensionCount() == 3:                
            #     info = QString("%1x%2x%3 cube")
            #             .arg(dataset.extents()[0])
            #             .arg(dataset.extents()[1])
            #             .arg(dataset.extents()[2])
            # else:
            #     info = QString("%1 dimensional object").arg(dataset.dimensionCount())

            node = ExdirTreeItem(row, 0, depth + 1,
                                                key,
                                                parent.path + "/" + key,
                                                type,
                                                parent)
            node.setInfo(info)

            if isinstance(object, exdir.core.Group):            
                node.needsChildIteration = True

            row += 1



    def loadFile(self):
        if not self.source.isValid() or self.source.isEmpty():        
            print("Not loading because", self.source)
            return

        filePath = self.source.toLocalFile()
        filenameOnly = filePath
        fileInfo = QFileInfo(filePath)
        if fileInfo.exists():
            filenameOnly = fileInfo.fileName()

        file = exdir.File(filePath)
        self.root = ExdirTreeItem(0, 0, 0, "", "", "", None)
        self.beginInsertRows(QModelIndex(), 0, 0)
        fileItem = ExdirTreeItem(0, 0, 1, filenameOnly, "", "File", self.root)
        self.endInsertRows()
        self.addChildObjects(fileItem, file, 0)
        self.dataChanged.emit(QModelIndex(), QModelIndex())
        
    sourceChanged = pyqtSignal(QUrl)
    source = pyqtProperty(QUrl, source, setSource, notify=sourceChanged)

class ExdirTreeItem(QObject):
    def __init__(self, row_, column_, depth_, name_, path_, type, parent):
        super().__init__(parent)
        self.row = row_
        self.column = column_
        self.depth = depth_
        self._name = name_
        self._path = path_
        self._type = type
        self._info = ""
        self.parentItem = parent
        self.needsChildIteration = False
        
    def child(self, row):
        return self.children()[row]

    def name(self):
        return self._name

    def path(self):
        return self._path

    def type(self):
        return self._type

    def info(self):
        return self._info

    def setName(self, name):
        if self._name == name:
            return

        self._name = name
        self.nameChanged.emit()

    def setPath(self, path):
        if self._path == path:
            return

        self._path = path
        self.pathChanged.emit()

    def setType(self, type):
        if self._type == type:
            return

        self._type = type
        self.typeChanged.emit()

    def setInfo(self, info):
        if self._info == info:
            return

        self._info = info
        self.infoChanged.emit()

    nameChanged = pyqtSignal()
    pathChanged = pyqtSignal()
    typeChanged = pyqtSignal()
    infoChanged = pyqtSignal()
    name = pyqtProperty(str, name, setName, notify=nameChanged)
    path = pyqtProperty(str, path, setPath, notify=pathChanged)
    type = pyqtProperty(str, type, setType, notify=typeChanged)
    info = pyqtProperty(str, info, setInfo, notify=infoChanged)
